shell/common/engine_unittests.cc

```diff
@@ -418,4 +571,199 @@ TEST_F(EngineTest, PassesLoadDartDeferredLibraryErrorToRuntime) {
+    // metrics is not empty.
+    engine.AddView(kFlutterImplicitViewId, ViewportMetrics{1.0, 10, 10, 1, 0});
+  });
+  continuation_ready_latch.Wait();
+
+  auto configuration = RunConfiguration::InferFromSettings(settings_);
+  configuration.SetEntrypoint("renderWarmUpImplicitView");
+  engine_context->Run(std::move(configuration));
+
+  draw_latch.Wait();
+}
+
+// The warm up frame should work if only some of the registered views are
+// included.
+//
+// This test also verifies that the warm up frame can render multiple views.
+TEST_F(EngineTest, AnimatorSubmitPartialViewsForWarmUp) {
+  MockAnimatorDelegate animator_delegate;
+  std::unique_ptr<EngineContext> engine_context;
+
+  std::shared_ptr<PlatformMessageHandler> platform_message_handler =
+      std::make_shared<MockPlatformMessageHandler>();
+  EXPECT_CALL(delegate_, GetPlatformMessageHandler)
+      .WillOnce(ReturnRef(platform_message_handler));
+
+  fml::AutoResetWaitableEvent continuation_ready_latch;
+  fml::AutoResetWaitableEvent draw_latch;
+  EXPECT_CALL(animator_delegate, OnAnimatorDraw)
+      .WillOnce(
+          Invoke([&draw_latch](const std::shared_ptr<FramePipeline>& pipeline) {
+            auto status =
+                pipeline->Consume([&](std::unique_ptr<FrameItem> item) {
+                  EXPECT_EQ(item->layer_tree_tasks.size(), 2u);
+                  EXPECT_EQ(item->layer_tree_tasks[0]->view_id, 1);
+                  EXPECT_EQ(item->layer_tree_tasks[1]->view_id, 2);
+                });
+            EXPECT_EQ(status, PipelineConsumeResult::Done);
+            draw_latch.Signal();
+          }));
+  EXPECT_CALL(animator_delegate, OnAnimatorBeginFrame)
+      .WillRepeatedly(
+          Invoke([&engine_context, &continuation_ready_latch](
+                     fml::TimePoint frame_target_time, uint64_t frame_number) {
+            continuation_ready_latch.Signal();
+            engine_context->EngineTaskSync([&](Engine& engine) {
+              engine.BeginFrame(frame_target_time, frame_number);
+            });
+          }));
+
+  std::unique_ptr<Animator> animator;
+  PostSync(task_runners_.GetUITaskRunner(),
+           [&animator, &animator_delegate, &task_runners = task_runners_] {
+             animator = std::make_unique<Animator>(
+                 animator_delegate, task_runners,
+                 static_cast<std::unique_ptr<VsyncWaiter>>(
+                     std::make_unique<testing::ConstantFiringVsyncWaiter>(
+                         task_runners)));
+           });
+
+  engine_context = EngineContext::Create(delegate_, settings_, task_runners_,
+                                         std::move(animator));
+
+  engine_context->EngineTaskSync([](Engine& engine) {
+    // Schedule a frame to make the animator create a continuation.
+    engine.ScheduleFrame(true);
+    // Add multiple views.
+    engine.AddView(0, {1, 10, 10, 22, 0});
+    engine.AddView(1, {1, 10, 10, 22, 0});
+    engine.AddView(2, {1, 10, 10, 22, 0});
```

Comment:
Consider adding ViewportMetrics for readability:

```suggestion
    engine.AddView(0, ViewportMetrics{1, 10, 10, 22, 0});
    engine.AddView(1, ViewportMetrics{1, 10, 10, 22, 0});
    engine.AddView(2, ViewportMetrics{1, 10, 10, 22, 0});
```